vent/coordinator/control_module.py

```diff
@@ -221,8 +198,51 @@ def update(self, pressure):
             if self.pressure < self.PEEP:
                 self.Qin = 1
         else:
-            self.cycle_start = time.time() # restart
+            self.cycle_start    = time.time() # new cycle starts
+            self.cycle_counter += 1
+
+        if self.cycle_counter not in self.cycle_waveforms.keys():   #if this cycle doesn't exist yet, start it
+            self.cycle_waveforms[self.cycle_counter] = np.array([[0, pressure]])
+        else:
+            data = self.cycle_waveforms[self.cycle_counter]
+            data = np.append(data, [[cycle_phase, pressure]], axis=0)
+            self.cycle_waveforms[self.cycle_counter] = data
+
+def test_critical_levels(min, max, value, name, alarms) -> List[Alarm]:
+    '''
+    This tests whether a variable is within bounds.
+    If it is, and an alarm existed, then the "alarm_end_time" is set.
+    If it is NOT, a new alarm is generated and appendede to the alarm-list.
+    Input:
+        min:    minimum value  (e.g. 2)
+        max:    maximum value  (e.g. 5)
+        value:  test value   (e.g. 3)
+        name:   parameter type (e.g. "PIP", "PEEP" etc.)
+        alarms: list of existing alarms 
+    Returns:
+        alarms: updated list of existing alarms
+
+    '''
+    if (value<min) or (value>max):           # If the variable is not within limits, make sure there is an alarm.
+        new_alarm = Alarm(alarm_name = name, is_active = True, severity = AlarmSeverity.RED, \
```

Comment:
Creating the new alarm could be moved down into the 'if does_not_exist:' block

Reply:
thanks, done!